src/Microsoft.VisualStudio.ProjectSystem.Managed/Threading/Tasks/CancellationSeries.cs

```diff
@@ -0,0 +1,112 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+
+namespace Microsoft.VisualStudio.Threading.Tasks
+{
+    /// <summary>
+    /// Produces a series of <see cref="CancellationToken"/> objects such that requesting a new token
+    /// causes the previously issued token to be cancelled.
+    /// </summary>
+    /// <remarks>
+    /// <para>Consuming code is responsible for managing overlapping asynchronous operations.</para>
+    /// <para>This class has a lock-free implementation to minimise latency and contention.</para>
+    /// </remarks>
+    public sealed class CancellationSeries : IDisposable
+    {
```

Comment:
What are the chances that an instance of this class gets dropped without being Dispose'd?  Maybe add a debug-only finalizer to catch any such misuse, as that would potentially mean a registration from a linked token source was leaked.

Reply:
Current usages are long lived but it seems like a good idea to be able to catch this.

Is there any reason not to add a finalizer that does this, even in non-DEBUG builds?

Reply:
> Is there any reason not to add a finalizer that does this, even in non-DEBUG builds?

There's a small run-time cost to adding a finalizer, but it might be insignificant compared to everything else. The real issue would be what the finalizer does.... if it crashes, you probably wouldn't want that in release :)  If it just logs, and pops an assert in a debug build, should be fine.

Reply:
Fixed in eb02e72be2f150de51dfc12eea88ce4e539e8892. I went for an assert.